test/notifications/display_test.dart

```diff
@@ -79,6 +83,21 @@ void main() {
   TestZulipBinding.ensureInitialized();
   final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
 
+  makeFakeHttpClient({http.Response? response, Exception? exception}) => http_testing.MockClient((request) async {
+    assert((response != null) ^ (exception != null));
+    if (exception != null) throw exception;
+    return response!; // TODO return 404 on non avatar urls
+  });
+  final fakeHttpClient = makeFakeHttpClient(
+    response: http.Response.bytes(kSolidBlueAvatar, HttpStatus.ok));
+
+  T runWithHttpClient<T>(
+    T Function() callback, {
+    http.Client Function()? httpClientFactory,
+  }) {
+    return http.runWithClient(callback, httpClientFactory ?? () => fakeHttpClient);
```

Comment:
For reading the `?? () => fakeHttpClient`, I think it could help if `fakeHttpClient` had a more specific name, like `fakeHttpClientGivingSuccess` or something.